framework/calico_framework.py

```diff
@@ -0,0 +1,707 @@
+        if self.all_tests_complete:
+            self.report_results_and_exit()
+
+    def registered(self, driver, frameworkId, masterInfo):
+        """
+        Callback used when the framework is succesfully registered.
+        """
+        _log.info("REGISTERED: with framework ID %s", frameworkId.value)
+
+    def resourceOffers(self, driver, offers):
+        """
+        Triggered when the framework is offered resources by mesos.
+        """
+        for offer in offers:
+            new_offer = Offer(offer)
+            if new_offer.num_tasks_launchable == 0:
+                _log.info("REJECTING EMPTY OFFER: %s", new_offer)
+                driver.declineOffer(offer.id)
+            else:
+                _log.info("New: %s", new_offer)
+                self.offers.add(new_offer)
+        for offer in self.offers:
+            _log.debug(offer)
+
+        # First, hand all offers to the running tests
+        _log.info("Offering to Running Tests")
+        for test in [test for test in self.tests if test.running]:
+            _log.info("\t%s", test)
+            test.launch(self.offers)
+
+        # Now that some offers have been reserved, offer the remaining to the
+        # unstarted tests
+        _log.info("Offering to Unstarted Tests")
+        for test in [test for test in self.tests if
+                     not test.running and not test.complete]:
+            _log.info("\t%s", test)
+            if test.can_run_on([offer for offer in self.offers if offer.reserved_by in [None, test]]):
+                _log.info("\t\tAll Tasks Succesfully Planned. Launching")
+                test.launch(self.offers)
+
+        # If there's no running tests, start timeouts
+        if not [test for test in self.tests if test.running]:
+            for test in [test for test in self.tests if
+                         not test.running and not test.complete]:
+                if test.timed_out(start=True):
+                    self.kill_test(test, "Test timed out waiting for resources")
+        else:
+            # There is a running test. remove every other
+            # test's unstarted-test-timer
+            for test in [test for test in self.tests if not test.running]:
+                test.timeout = None
+
+        if self.all_tests_complete:
+            _log.info("All Tests Complete!")
+            self.report_results_and_exit()
+
+        _log.info("No Tests ready to launch more tasks.")
+
+    def report_results_and_exit(self, error=None):
+        if error:
+            _log.error("KILLING FRAMEWORK: %s", error)
+        for test in self.tests:
+            test.print_report()
+            print "\n"
+        driver.stop()
+
+    def statusUpdate(self, driver, update):
+        """
+        Triggered when the Framework receives a task Status Update from the
+        Executor
+        """
+        # Find the task which corresponds to the status update
+        try:
+            calico_task = next(task for task in self.all_tasks if
```

Comment:
You need to think about frequency not in terms of number of appearances in the code, but the number of times the code will be executed.  In this case, every time we get a task update.

It's true that `next()` only calls the iterator once, but it depends on the implementation details how expensive it will be to create the iterator.  I looked into this a little more---the expression as you've written is is implemeted as a "lazy" generator, but the `self.all_tasks` is actually a property which allocates a new list and copies from all the tests.  So, while the comparison against task ID will terminate when it finds a match, you've actually already looped through all the tasks once to create `self.all_tasks`.